### PR TITLE
[TACHYON-523] Implement PartialLRUEvictor

### PR DESCRIPTION
**Pull request by Yours-shupeng**
Reopen PR [1034](https://github.com/amplab/tachyon/pull/1034)

### CODE REVIEW

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java [hunk 1]
```diff
@@ -30,6 +30,10 @@
    */
   LRU,
   /**
+   * Evict old blocks among several StorageDirs by PartialLRU
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: Better to have a brief description here to tell the difference between LRU and PartialLRU.

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+import tachyon.Constants;
+import tachyon.Pair;
+import tachyon.worker.block.BlockMetadataManager;
+import tachyon.worker.block.BlockStoreEventListenerBase;
+import tachyon.worker.block.BlockStoreLocation;
+import tachyon.worker.block.meta.BlockMeta;
+import tachyon.worker.block.meta.StorageDir;
+import tachyon.worker.block.meta.StorageTier;
+
+public class PartialLRUEvictor extends BlockStoreEventListenerBase implements Evictor {
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: Please javadoc this class to explain what this Evictor is, especially what PartialLRU is given people usually understand what LRU is.

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+              + mMeta.getBlockMeta(blockId).getBlockSize());
+        } else {
+          pendingBytesInDir.put(toDir, mMeta.getBlockMeta(blockId).getBlockSize());
+        }
+      }
+    }
+    return new EvictionPlan(toMove, toEvict);
+  }
+
+  public Pair<StorageDir, List<Long>> getDirCandidate(long availableBytes,
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: (1) Why this method is `public`? Is it supposed to be used by external classes? If not, make it `private`
(2)  Could you also javadoc this method?

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        ignoreList.add(selectedDir);
+        selectedDir = getDirWithMaxFreeSpace(availableBytes, location, ignoreList);
+      } else {
+        return new Pair<StorageDir, List<Long>>(selectedDir, victimBlocks);
+      }
+    }
+    return null;
+  }
+
+  public StorageDir getDirWithMaxFreeSpace(long availableBytes, BlockStoreLocation location,
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: Same comment with respect to `public` and javadoc

**Comment by apc999** [comment]
A few comments left. Thanks for the patch!

**Comment by apc999** [comment]
+1 LGTM

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java [hunk 5]
```diff
@@ -30,6 +30,13 @@
    */
   LRU,
   /**
+   * Evict old blocks in certain StorageDir by LRU. The main difference between PartialLRU and
+   * LRU is that LRU choose old blocks among several StorageDirs until one StorageDir
+   * satisfies the request space, but PartialLRU select one StorageDir first and evict old
+   * blocks in certain StorageDir by LRU
+   */
+  PARTIAL_LRU,
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: Is this the same as the old Partial LRU evictor? I thought it was still evict by LRU but only evict from the final directory that is chosen.

Comment (Yours-shupeng) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @calvinjia Sorry for that I have made a mistake in the `getDirCandidate()` function in `PartialLRUEvictor.java`. The blocks should be evicted by LRU only if it exists in the selected `StorageDir`, but I evict in the whole `BlockStoreLocation`. However,  I have to scan the whole blocks to decide the LRU blocks to evict in the selected `StorageDir` or should I matain LRU blocks in each `StorageDir`?

Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: Either way can work, one algorithm could be to scan through all the blocks by LRU and bucket the blocks by `StorageDir` (ie. with a hashmap) until one `StorageDir` has enough space, then you can evict all the blocks in that bucket.

Comment (Yours-shupeng) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @calvinjia This is a little like the `LRUEvictor`. So should we choose a `StorageDir` with max free space first, and then scan all the blocks by LRU and bucket the blocks of the selected `StorageDir` until the selected `StorageDir` has enough space?

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @Yours-shupeng hi, shupeng, I haven't read the code, through the description of javadoc, the logic of PARTIAL_LRU is the same when passing a determined BlockStoreLocation(other than anyTier or anyDirInTier) to current LRUEvictor?

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+ * first and evict old blocks in certain StorageDir by LRU
+ */
+public class PartialLRUEvictor extends BlockStoreEventListenerBase implements Evictor {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private final BlockMetadataManager mMeta;
+
+  private static final int LINKED_HASH_MAP_INIT_CAPACITY = 200;
+  private static final float LINKED_HASH_MAP_INIT_LOAD_FACTOR = 0.75f;
+  private static final boolean LINKED_HASH_MAP_ACCESS_ORDERED = true;
+  private static final boolean UNUSED_MAP_VALUE = true;
```
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: could be prettier if we put all `private static final xxx` together.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java (hunk 5 above)
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @Yours-shupeng oh, I understand it through codes, so, PartialLRU means we first select StorageDir with the maximum possible to be available bytes after eviction, and then select blocks in this dir with LRU.

Comment (Yours-shupeng) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @uronce-cc Yes, when the `BlockStoreLocation` is some determined `StorageDir`, `PARTIAL_LRU` is the same to `LRUEvictor`. `PARTIAL_LRU` is just a little logic changes to `LRUEvictor` overall.

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java (hunk 6 above)
Comment (Yours-shupeng) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: @uronce-cc Thanks for the comment, I wil fix it.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java (hunk 5 above)
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @Yours-shupeng yeah, maybe we can share codes between `PartialLRUEvictor` and `LRUEvictor`, if we let `PartialLRUEvictor extends LRUEvictor`, then after selecting the `StorageDir`, we can just call `super` in `freeSpace`.  then PartialLRUEvictor will have feature like "cascading eviction"(https://github.com/amplab/tachyon/pull/1079)  once LRUEvictor has it. What do you think about?

Comment (Yours-shupeng) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @uronce-cc Thanks for the advice, I think it's nice to let `PartialLRUEvictor` extends `LRUEvictor`. But there is still a problem now, I have to override `freeSpace(long, BlockStoreLocation, EvictionPlan)` to support select a `StorageDir` first after "cascading eviction" is merged in `master`. So should I let `PartialLRUEvictor` extends `LRUEvictor` now and adjust later or do you have any other suggestions?

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @Yours-shupeng yup, let's postpone this until LRUEvictor is good to ship.

**Comment by haoyuan** [comment]
We can’t automatically merge this pull request.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java (hunk 5 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @Yours-shupeng Is the current logic the same as the previous Partial LRU Evictor?

Comment (Yours-shupeng) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @calvinjia I think this is logic of LRU Evictor. LRU Evictor simulate LRU eviction in each directory until a dir has enough space. I have referred the code of [branch-0.6](https://github.com/amplab/tachyon/blob/branch-0.6/core/src/main/java/tachyon/worker/eviction/EvictPartialLRU.java), I think the Partial LRU Evictor mainly focus on choosing dir with max free space to ensure the uniform distribution of blocks.

Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: @Yours-shupeng Oh I see, thanks for looking into it!

**Comment by calvinjia** [comment]
@Yours-shupeng The code looks good to me, could you add unit tests for the specific eviction strategy of this evictor?

**Comment by apc999** [comment]
@apc999 @calvinjia  please review this PR

**Comment by haoyuan** [comment]
We can’t automatically merge this pull request.

**Comment by Yours-shupeng** [comment]
@calvinjia I'm sorry to be late, unit tests are added now.

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+import tachyon.worker.block.meta.BlockMeta;
+import tachyon.worker.block.meta.StorageDir;
+import tachyon.worker.block.meta.StorageDirView;
+import tachyon.worker.block.meta.StorageTier;
+import tachyon.worker.block.meta.StorageTierView;
+
+/**
+ * This class is used to evict old blocks in certain StorageDir by LRU. The main difference
+ * between PartialLRU and LRU is that LRU choose old blocks among several StorageDirs
+ * until one StorageDir satisfies the request space, but PartialLRU select one StorageDir
```
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: What about specifying "select one StorageDir with maximum free space"?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java [hunk 8]
```diff
@@ -30,6 +30,13 @@
    */
   LRU,
   /**
+   * Evict old blocks in certain StorageDir by LRU. The main difference between PartialLRU and
+   * LRU is that LRU choose old blocks among several StorageDirs until one StorageDir
+   * satisfies the request space, but PartialLRU select one StorageDir first and evict old
```
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorType.java]: What about specifying "select one StorageDir with maximum free space"?

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    // 2. iterate over blocks in LRU order until the candidate StorageDir can satisfy
+    // bytesToBeAvailable after evicting its blocks iterated so far
+    List<Long> candidateBlocks = new ArrayList<Long>();
+    long freedBytes = 0;
+    Iterator<Map.Entry<Long, Boolean>> it = mLRUCache.entrySet().iterator();
+    while (it.hasNext() && candidateDirView.getAvailableBytes() + freedBytes < bytesToBeAvailable) {
+      long blockId = it.next().getKey();
+      try {
+        BlockMeta block = mManagerView.getBlockMeta(blockId);
+        if (null != block) { // might not present in this view
```
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: if the block is not present, `NotFoundException` will be thrown?

File: servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  @Override
+  protected StorageDirView cascadingEvict(long bytesToBeAvailable, BlockStoreLocation location,
+      EvictionPlan plan) {
+
+    StorageDirView candidateDirView = null;
+    // 1. Get StorageDir with max free space. If no such StorageDir, return null. If
+    // bytesToBeAvailable can already be satisfied without eviction, return emtpy plan
+    candidateDirView = getDirWithMaxFreeSpace(bytesToBeAvailable, location);
```
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/PartialLRUEvictor.java]: Above two lines can be combined, no need for explicit initialization to `null`.

**Comment by Yours-shupeng** [comment]
@uronce-cc Thanks for the comments. I have fixed them.

**Comment by uronce-cc** [comment]
+1 LGTM

**Comment by calvinjia** [comment]
+1 LGTM

**Pull request merged**